internal/controller/stas/workload_controller.go

```diff
@@ -171,52 +170,54 @@ func (r *PodReconciler) reconcile(ctx context.Context, pod *corev1.Pod) error {
-		cis.Name, err = imageScanName(podController, containerName, image.Image)
+		name, err := imageScanName(podController, containerName, image.Image)
 		if err != nil {
 			return err
 		}
 
-		mutateFn := func() error {
-			cis.Labels = pod.GetLabels()
-			cis.Spec.Workload.Group = podController.GetObjectKind().GroupVersionKind().Group
-			cis.Spec.Workload.Kind = podController.GetObjectKind().GroupVersionKind().Kind
-			cis.Spec.Workload.Name = podController.GetName()
-			cis.Spec.Workload.ContainerName = containerName
-			cis.Spec.Image = image.Image
-			cis.Spec.Tag = image.Tag
-			// Ensure MinSeverity unset until we eventually make use of it
-			cis.Spec.MinSeverity = nil
-
-			if v := podController.GetAnnotations()[stasv1alpha1.WorkloadAnnotationKeyIgnoreUnfixed]; v == "true" {
-				cis.Spec.IgnoreUnfixed = ptr.To(true)
-			} else {
-				cis.Spec.IgnoreUnfixed = ptr.To(false)
-			}
-
-			owners := getCISOwners(containerName, image)
-			if len(owners) == 0 {
-				// Safeguard to validate assumption in `cisOwnerLookup`.
-				return errors.New("Found no owners for CIS")
-			}
+		cisObj := &stasv1alpha1.ContainerImageScan{}
+		cisObj.Namespace = pod.Namespace
+		cisObj.Name = name
+
+		cis := stasv1alpha1ac.ContainerImageScan(name, pod.Namespace).
+			WithLabels(pod.GetLabels()).
+			WithSpec(
+				stasv1alpha1ac.ContainerImageScanSpec().
+					WithWorkload(
+						stasv1alpha1ac.Workload().
+							WithGroup(podController.GetObjectKind().GroupVersionKind().Group).
+							WithKind(podController.GetObjectKind().GroupVersionKind().Kind).
+							WithName(podController.GetName()).
+							WithContainerName(containerName),
+					).
+					WithName(image.Image.Name).
+					WithDigest(image.Image.Digest).
+					WithTag(image.Tag).
+					WithIgnoreUnfixed(podController.GetAnnotations()[stasv1alpha1.WorkloadAnnotationKeyIgnoreUnfixed] == "true"),
+			)
+
+		owners := getCISOwners(containerName, image)
+		if len(owners) == 0 {
+			// Safeguard to validate assumption in `cisOwnerLookup`.
+			return errors.New("Found no owners for CIS")
+		}
 
-			for _, owner := range owners {
-				if err := controllerutil.SetOwnerReference(&owner, cis, r.Scheme); err != nil {
-					return err
-				}
+		for _, owner := range owners {
+			if err := SetOwnerReference(&owner, cis.ObjectMetaApplyConfiguration, r.Scheme); err != nil {
+				return err
 			}
+		}
 
-			return nil
+		// Deep copy of object to avoid pollution with managed fields not accepted by the following patch.
+		if err := upgradeManagedFields(ctx, r.Client, cisObj.DeepCopyObject().(client.Object), fieldOwner); err != nil {
+			return err
 		}
 
-		_, err = controllerutil.CreateOrUpdate(ctx, r.Client, cis, mutateFn)
-		if err != nil {
+		if err := r.Patch(ctx, cisObj, applyPatch{cis}, FieldValidationStrict, client.ForceOwnership, fieldOwner); err != nil {
 			return err
 		}
 
-		err = r.garbageCollectObsoleteImageScans(ctx, pod, cis)
+		err = r.garbageCollectObsoleteImageScans(ctx, pod, name, containerName)
```

Comment:
```suggestion
		err = r.garbageCollectObsoleteImageScans(ctx, pod, cisObj)
```

To avoid changing this in this PR.

Reply:
Resoning here is that `upgradeManagedFields` will fetch the resource and populate the required field? (`.Spec.Workload.ContainerName`)